File: Python/Zadanie 2/szachy_python/pieces/bishop.py
from piece import Piece

class Bishop(Piece):
    def symbol(self):
        return '♗' if self.color == 'white' else '♝'

    def is_valid_move(self, board, start, end):
        dx = end[0] - start[0]
        dy = end[1] - start[1]
        if abs(dx) != abs(dy):
            return False

        step_x = 1 if dx > 0 else -1
        step_y = 1 if dy > 0 else -1

        x, y = start[0] + step_x, start[1] + step_y
        while (x, y) != end:
            if board.get_piece((x, y)) is not None:
                return False
            x += step_x
            y += step_y

        target = board.get_piece(end)
        return target is None or target.color != self.color

File: C#/Zadanie 2/Szachy/Bishop.cs
namespace Szachy;

public class Bishop : Piece
{
    public Bishop(PieceColor color) : base(color) {}

    public override char Symbol => Color == PieceColor.White ? 'B' : 'b';

    public override bool IsValidMove((int x, int y) from, (int x, int y) to, Board board)
    {
        int dx = Math.Abs(to.x - from.x);
        int dy = Math.Abs(to.y - from.y);

        if (dx != dy) return false;  // tylko ruchy po skosie

        int stepX = (to.x - from.x) / dx;
        int stepY = (to.y - from.y) / dy;

        int x = from.x + stepX;
        int y = from.y + stepY;

        // Sprawdź, czy na drodze nie ma żadnych figur
        while (x != to.x && y != to.y)
        {
            if (board.Squares[x, y] != null)
                return false;

            x += stepX;
            y += stepY;
        }

        // Sprawdź, czy na polu docelowym nie stoi własna figura
        var destinationPiece = board.Squares[to.x, to.y];
        if (destinationPiece != null && destinationPiece.Color == this.Color)
            return false;

        return true;
    }
}